lib/util/result.ts

```diff
@@ -1,73 +1,293 @@
-  static wrap<T>(
-    input: (() => T) | Promise<T>
-  ): Result<T> | Promise<Result<T>> {
-    return input instanceof Promise
-      ? Result.wrapPromise(input)
-      : Result.wrapCallback(input);
+  transform<U, EE>(
+    fn: (value: NonNullable<T>) => Result<U, EE>
+  ): Result<U, E | EE>;
+  transform<U, EE>(
+    fn: (value: NonNullable<T>) => Promise<Result<U, EE>>
+  ): AsyncResult<U, E | EE>;
+  transform<U>(
+    fn: (value: NonNullable<T>) => Promise<NonNullable<U>>
+  ): AsyncResult<U, E | Error>;
+  transform<U>(
+    fn: (value: NonNullable<T>) => NonNullable<U>
+  ): Result<U, E | Error>;
+  transform<U, EE>(
+    fn: (
+      value: NonNullable<T>
+    ) =>
+      | NonNullable<U>
+      | Result<U, EE>
+      | Result<U, EE>
+      | Promise<NonNullable<U>>
+  ): Result<U, E | EE> | AsyncResult<U, E | EE> {
+    if (!this.res.ok) {
+      return Result.err(this.res.error);
+    }
+
+    try {
+      const res = fn(this.res.value);
+
+      if (res instanceof Promise) {
+        return new AsyncResult((resolve) => {
+          res
+            .then((newResult) =>
+              newResult instanceof Result
+                ? resolve(newResult)
+                : resolve(Result.ok(newResult))
+            )
+            .catch((error) => {
+              logger.warn(
+                { err: error },
+                'Result: unhandled async transform error'
+              );
+              resolve(Result.err(error) as never);
+            });
+        });
+      }
+
+      if (res instanceof Result) {
+        return res;
+      }
+
+      return Result.ok(res);
+    } catch (error) {
+      logger.warn({ err: error }, 'Result: unhandled transform error');
+      return Result.err(error);
+    }
   }
+}
 
-  private constructor(public readonly res: Res<T, E>) {}
+export class AsyncResult<T, E> extends Promise<Result<T, E>> {
+  constructor(
+    executor: (
+      resolve: (value: Result<T, E> | PromiseLike<Result<T, E>>) => void,
+      reject: (reason?: unknown) => void
+    ) => void
+  ) {
+    super(executor);
+  }
+
+  static wrap<T, E = Error>(promise: Promise<Result<T, E>>): AsyncResult<T, E>;
+  static wrap<T, E = Error>(
+    promise: Promise<NonNullable<T>>
+  ): AsyncResult<T, E>;
+  static wrap<T, E = Error>(
+    promise: Promise<NonNullable<T> | Result<T, E>>
+  ): AsyncResult<T, E> {
+    return new AsyncResult((resolve) => {
+      promise
+        .then((value) => {
+          if (value instanceof Result) {
+            return resolve(value);
+          }
 
-  transform<U>(fn: (value: T) => U): Result<U, E> {
-    return this.res.success
-      ? Result.ok(fn(this.res.value))
-      : Result.err(this.res.error);
+          return resolve(Result.ok(value));
+        })
+        .catch((error) => resolve(Result.err(error)));
+    });
   }
 
-  catch<U>(fallback: U): T | U {
-    return this.res.success ? this.res.value : fallback;
+  static wrapNullable<T, E, NullError, UndefinedError>(
```

Comment:
this needs tsdoc's to describe usage, it's not really clear to me for what it's useful. code looks also pretty complicated.